### ISSUE TITLE
How to do customized BOA install?

### ISSUE THREAD

**Issue by ar-jan** [omega8cc/boa#597]
Before BOA 2.4 I could modify the variables in BARRACUDA.sh.txt and OCTOPUS.sh.txt and then install by doing `bash BARRACUDA.sh.txt && bash OCTOPUS.sh.txt`, allowing to configure all desired settings in one go.

Is there an equivalent to doing this in BOA >= 2.4?

**Comment by omega8cc** [comment]
Config templates are there for years: https://github.com/omega8cc/boa/tree/master/docs/cnf

**Comment by ar-jan** [comment]
I saw the templates, but how to use them? Can I just put them in the /root dir?

I didn't find docs for how to use them (on initial install).

**Comment by omega8cc** [comment]
These are templates with built-in docs you should read.

You shouldn't run BARRACUDA directly for a long time already, you should use wrappers, as explained in the docs. Now it is finally no longer even possible.

You could also start installation and abort it once it creates these config files for you, re-configure them and run installation again.

But why not to do that on upgrade and not initial install?

It is always recommended to run upgrade after initial install before using the system in production.

**Comment by ar-jan** [comment]
I'm familiar with the recommended install procedure via `boa in-stable`, but my question is about whether or not there are further options. There are some things that seemed to hint at this.

For example there are variables like `_SMTP_RELAY_HOST` that can only be configured on initial install, but there is no (documented) way how one could use these when using `boa in-stable`. So when I used BARRACUDA.sh.txt directly I assumed I was using an undocumented feature allowing to define more settings on initial install.

I'm happy to use `boa in-stable` (especially now that we can define PHP version as well), but I like to understand the possibilities.

**Comment by ar-jan** [comment]
Also, of course, thanks for the awesome new release! :)

**Comment by omega8cc** [comment]
All important options can be configured via configuration file created prior to running the initial installation. There should be no need to configure this, though, unless you have some special system requirements, like blocked port 25, which is very rare these days. Everything else can be configured before you run upgrade, which is recommended after initial install. We have made it impossible to edit and run scripts directly to avoid many hard to track and fix issues. This should make things easier for everyone. Thanks!

**Comment by ar-jan** [comment]
Makes sense. I have one question left: how to utilize the `configuration file created prior to running the initial installation` for the more advanced use case. As far as I can see this isn't documented?

The _SMTP_RELAY_HOST is actually just what I've used this for previously on RunAbove, since they do block port 25 (at least for the time being).

**Comment by omega8cc** [comment]
Please check the docs here:

https://github.com/omega8cc/boa/blob/master/docs/INSTALL.txt#L29
https://github.com/omega8cc/boa/blob/master/docs/NOTES.txt#L4
https://github.com/omega8cc/boa/blob/master/docs/NOTES.txt#L39
https://github.com/omega8cc/boa/blob/master/docs/NOTES.txt#L94

**Comment by ar-jan** [comment]
The places you link to tell me: 1) The configuration should be located at /root/.barracuda.cnf 2) There is a template with example options and their explanations 3) To customize your configuration you should install boa as per INSTALL.txt and then customize the auto-created /root/.barracuda.cnf + upgrade.

The above customize-on-upgrade was clear to me, but what still needs to be addressed, I think:
If you want to install BOA with a "configuration file created prior to running the initial installation", e.g. in order to configure some of the initial-install-only settings like _SMTP_RELAY_HOST, then:
- Should you place a customized /root/.barracuda.cnf with all the settings from the template, or perhaps only the initial-install-only settings?
- What command do you run? `boa in-stable`  without arguments? Or should you provide the required parameters as usual?

**Comment by omega8cc** [comment]
Feel free to propose a patch to extend the docs and explain advanced topics like this.

**Comment by ar-jan** [comment]
Sure, but those two points are actually still questions for me. Well, I assume you need the whole template as /root/.barracuda.cnf, but what should be the installation command?

**Comment by omega8cc** [comment]
The same commands, templates used as explained in their built-in docs. No magic involved! :)

**Comment by AntiNSA** [comment]
I am curious about this too. I never need the CHV option, and always need the SR4 option. Also, I have to change my port on initial install. The only two ways I know to do this is either ctrl z after the barracussa.cnf is created, dfelete the pid's, and then re run the process... or to do the update again after already waiting for the initial install.

**Comment by ar-jan** [comment]
Here's my understanding so far:

- It is strongly recommended to customize your installation on upgrade as already documented. Only if you need to install using settings for initial installation only (such as SMTP relay) should you create the /root/.barracuda.cnf file before initial install.
- If you do use the customized install, you should create /root/.barracuda.cnf with all settings, and still use the regular install command with the required options, like `boa in-stable public boa.example.com info@example.com o1`.
- I believe the .o1.octopus.cnf is not strictly required if you do customized install, since you need to provide the octopus name anyway to the `boa in-stable` command, but if you leave it out, you must at least be using php-55 because this is the default that would be used. But I think if you do use the customized install it's better to provide both .barracuda.cnf and .o1.octopus.cnf.
- I don't know which settings would take precedence, e.g. if you provide another hostname in your config file than what you specify in the `boa in-stable` command. But you wouldn't do that anyway, right? :)
- It's impossible to specify _STRONG_PASSWORDS=YES on initial install, you must do this on upgrade because `randpass` is not available when you've only run the meta-installer (see #596). Or you should manually add `export PATH=$PATH:/usr/local/bin:/opt/local/bin` to your /root/.bashrc after you've run the meta-installer, and log out and back in before you start the installation.

**Comment by BrianGilbert** [comment]
@AntiNSA You can download barracuda.cnf from this repo to /root/barracuda.cnf before running the install and populate the values you want.

**Comment by arbelaezjm** [comment]
Hello . After the initial instalation is there any problem involved with changing the port 22 on the .barracuda.cnf   , because i am reaciving to much ip blocked email from the firewall beacause and attack . i think if i change the port the posibilities to be attacked are less..

**Comment by ar-jan** [comment]
You can configure the `_SSH_PORT` variable in .barracuda.cnf to make sure the port number persists when upgrading BOA. In order for the changes to take effect directly, you should also change the port in /etc/ssh/sshd_config, and do `service ssh restart`.
